File: structures/trie.cpp
#include <bits/stdc++.h>

using namespace std;

class Trie{
private:
	class TrieNode{
		vector<TrieNode*> childs;
		bool endWord;
		TrieNode(){
			childs = vector<TrieNode*> (26,0);
			endWord = false;
		}
	};

	TrieNode *root;
public:
	Trie(){
		root = new TrieNode();
	}

	void insert(const string &word){
		int index = 0;
		TrieNode* it = root;

		while(index<word.size()){
			int pos = word[index]-'a';
			if(!it->childs[pos])
				it->childs[pos] = new TrieNode();

			it = it->childs[pos];
		}

		it->endWord = true;
	}

	bool query(const string &word){
		int index = 0;
		TrieNode *it = root;

		while(index<word.size()){
			int pos = word[index]-'a';
			if(!it->childs[pos])
				return false;
			it = it->childs[pos];
		}

		return it->endWord;
	}

	void printWords(){
			string temp;
			TrieNode *it = root;
			for(int i=0;i<it->childs.size();i++){
				if(it->childs[i]){
					temp.push_back(i+'a');
					printWord(temp,it->childs[i]);
					temp.pop_back();
				}
			}
	}

	void printWord(string word,TrieNode *node){
		if(node->endWord)
			cout<<word<<endl;

		for(int i=0;i<node->childs.size();i++){
			if(node->childs[i]){
				word.push_back(i+'a');
				printWord(temp,node->childs[i]);
				word.pop_back();
			}
		}
	}

};

int main(){

	Trie t;
	string s;

	while(cin>>s){
		t.insert(s);
	}

	t.printWords();
}

File: Kotlin/src/main/kotlin/org/algorithm/structures/Trie.kt
package org.algorithm.structures

class Trie {
    data class TrieNode(
        var childs: Array<TrieNode?> = arrayOfNulls<TrieNode?>(26),
        var endWord: Boolean = false
    )

    var root: TrieNode?

    init {
        root = TrieNode()
    }

    fun addWord(word: String) {
        var index = 0;
        var iterator: TrieNode? = root;
        while (index < word.length) {
            val pos = word[index] - 'a';
            if (iterator!!.childs[pos] == null) {
                iterator.childs[pos] = TrieNode();
            }
            iterator = iterator.childs[pos];
            index++;
        }

        iterator?.endWord = true;
    }

    fun queryWord(word: String): Boolean {
        var iterator: TrieNode? = root;
        for (letter in word) {
            val pos = letter - 'a';
            if (iterator!!.childs[pos] == null)
                return false;
            iterator = iterator.childs[pos];
        }
        return iterator!!.endWord;
    }
}